misc/docerina/src/main/java/org/ballerinalang/docgen/docs/BallerinaDocGenerator.java

```diff
@@ -226,23 +262,98 @@ private static void writeAPIDocs(ModuleLibrary moduleLib, Path output, boolean i
     }
 
     private static void copyDocerinaUI(Path output) {
-        File source = Path.of(System.getProperty("ballerina.home"), "lib", "tools", "doc-ui").toFile();
-        File dest;
-        if (source.exists()) {
-            dest = output.toFile();
-            try {
-                FileUtils.copyDirectory(source, dest);
-            } catch (IOException e) {
-                log.error("Failed to copy the doc ui.", e);
+        String source = RepoUtils.getRemoteRepoURL();
+        Path docsDirPath = ProjectUtils.createAndGetHomeReposPath().resolve(DOCS_FOLDER_NAME);
+        Path sha256FilePath = docsDirPath.resolve(SHA256_HASH_FILE_NAME);
+        Path zipFilePath = docsDirPath.resolve(BALLERINA_DOC_UI_ZIP_FILE_NAME);
+        OkHttpClient client = new OkHttpClient();
+        Request request = new Request.Builder()
+                .get()
+                .url(source)
+                .build();
+        try (Response response = client.newCall(request).execute()) {
+            if (response.code() == HTTP_OK && response.body() != null) {
+                ResponseBody responseBody = response.body();
+                JsonObject jsonResponse = JsonParser.parseReader(responseBody.charStream()).getAsJsonObject();
+                String sha256HashValue = jsonResponse.get(JSON_KEY_HASH_VALUE).getAsString();
+                String zipFileURL = jsonResponse.get(JSON_KEY_FILE_URL).getAsString();
+                if (!Files.exists(sha256FilePath) || !Files.exists(zipFilePath)) {
+                    if (!docsDirPath.toFile().exists()) {
+                        Files.createDirectories(docsDirPath);
+                    }
+                    writeFileInCache(zipFileURL, sha256HashValue, zipFilePath, sha256FilePath);
+                } else {
+                    String hashValueInCache = Files.readString(sha256FilePath).trim();
+                    if (!sha256HashValue.equals(hashValueInCache)) {
+                        writeFileInCache(zipFileURL, sha256HashValue, zipFilePath, sha256FilePath);
+                    }
+                }
+                copyDocUIToProjectDir(output, zipFilePath);
+            } else {
```

Comment:
We can use `early return pattern` and reduce the nesting levels here. https://medium.com/swlh/return-early-pattern-3d18a41bba8. You can check the negative of L275 and throw an error early and continue the rest.

Reply:
Fixed in https://github.com/ballerina-platform/ballerina-lang/pull/42386/commits/fff5836b373e5e37d3680640581edf928a6344c0